backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/VdsManager.java

```diff
@@ -1293,11 +1333,40 @@ public void updateV2VJobInfos(List<V2VJobInfo> v2vJobInfos) {
         }
     }
 
+
     private void initAvailableCpus() {
         List<VdsCpuUnit> cpuTopology = cachedVds.getCpuTopology();
-        if (cpuTopology == null) {
+        if (cpuTopology == null || cachedVds.getVdsmCpusAffinity() == null) {
             return;
         }
         this.cpuTopology = cpuTopology;
+        this.cpuTopology.stream().filter(cpu -> cpu.getCpu() == Integer.parseInt(cachedVds.getVdsmCpusAffinity())).forEach(cpu -> cpu.setVmId(Guid.SYSTEM));
+    }
+
+    public List<VdsCpuUnit> getCpuTopology() {
+        List<VdsCpuUnit> clone = new ArrayList<>();
+        cpuTopology.forEach(cpu -> clone.add(cpu.clone()));
+        return clone;
+    }
+
+    public void unpinVmCpus(VM vm) {
+        String pinning = vm.getCpuPinningPolicy() == CpuPinningPolicy.MANUAL ? vm.getCpuPinning() : vm.getCurrentCpuPinning();
+        if (pinning != null && !pinning.isBlank()) {
+            synchronized (this) {
```

Comment:
about the synchronization, the scheduling is done sequentially, right? are we concerned about a case that the ProcessDownVm would be called at the same time as another process that unpin the vCPUs?

Reply:
Yes. Generally I also want the cached topology to be safe and up-to-date.

Reply:
ok but that would mean that if we have two threads that execute ProcessDownVm, they might block each other even when the vms are pinned to completely different pcpus. it would be better to reduce the scope of the lock in a similar way to how it's done on Java concurrent collections where they don't lock the entire collections but only parts of it that are relevant